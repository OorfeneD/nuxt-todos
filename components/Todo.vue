<template>
  <li :class="`border border-gray-100 text-gray-500 rounded-md my-1 pb-2 ${todo.status ? 'bg-gray-100' : ''}`">
    <div class="pt-2 px-3 flex">
      <input
        :id="todo.id"
        class="flex-grow-0 cursor-pointer"
        :checked="todo.status"
        type="checkbox"
        name="todo"
        @change="$emit('updatetodo', { todo, status: $event.target.checked, originalEvent: $event})"
      >
      <label class="flex-grow text-center pl-1 pr-3 cursor-pointer" :for="todo.id">
        <strike v-if="todo.status" class="text-gray-500 transition transition-all">
          {{ todo.content.title }}
        </strike>
        <span v-else>
          {{ todo.content.title }}
        </span>
      </label>
      <button class="font-bold text-red-800 flex-grow-0 -my-2 -mx-3 px-3 hover:bg-red-400 hover:text-white transition rounded-r" @click="$emit('removetodo', { todo: todo, originalEvent: $event })">
        <icon close />
      </button>
    </div>
    <div v-if="!!todo.tags" class="flex flex-row px-1 pt-1">
      <tag-list class="pl-2" :tags="tags.filter(v => todo.tags.includes(v.id))" />
    </div>
  </li>
</template>

<script>
import TagList from './tags/TagList.vue'
export default {
  components: { TagList },
  props: {
    todo: {
      type: Object,
      default: () => ({
        id: -1,
        content: {
          title: '<i>dead inside</i>'
        }
      })
    },
    tags: {
      type: Array,
      default: () => []
    }
  }
}
</script>

<style>

</style>
