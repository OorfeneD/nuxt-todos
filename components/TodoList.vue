<template>
  <ul class="py-1">
    <todo
      v-for="todo in todos"
      :key="todo.id"
      :todo="todo"
      @removetodo="$emit('removetodo', $event)"
      @updatetodo="$emit('updatetodo', $event)"
    />
    <li v-if="todos.length < 1" class="text-sm text-gray-400 text-center py-2">
      There is nothing here. Add your first ToDo
    </li>
  </ul>
</template>

<script>
import { mapGetters } from 'vuex'
import Todo from './Todo.vue'
export default {
  components: { Todo },
  computed: {
    ...mapGetters(['todos'])
  }
}
</script>

<style>

</style>
