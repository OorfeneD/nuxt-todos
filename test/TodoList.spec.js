import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import TodoList from '../components/TodoList.vue'
import Todo from '../components/Todo'
import { todos } from './mocks/todo.mock'
const localVue = createLocalVue()
localVue.use(Vuex)

describe('TodoList Component', () => {
  let getters, emptyGetters, store, emptyStore
  let wrapper
  beforeEach(() => {
    getters = {
      todos: () => todos
    }
    emptyGetters = {
      todos: () => []
    }
    // eslint-disable-next-line
    store = new Vuex.Store({
      getters
    })
    // eslint-disable-next-line
    emptyStore = new Vuex.Store({
      getters: emptyGetters
    })
  })
  test('is a Vue instance', () => {
    wrapper = shallowMount(TodoList, { store, localVue })
    expect(wrapper.vm).toBeTruthy()
  })
  test('matchs snapshot', () => {
    wrapper = shallowMount(TodoList, { store, localVue })
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('renders with no todos', () => {
    wrapper = shallowMount(TodoList, { store: emptyStore, localVue })
    expect(wrapper.vm).toBeTruthy()
  })
  test('matchs no todos snapshot', () => {
    wrapper = shallowMount(TodoList, { store: emptyStore, localVue })
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('has message about empty todos', () => {
    wrapper = shallowMount(TodoList, { store: emptyStore, localVue })
    expect(wrapper.text()).toContain('Add your first ToDo')
  })
  test('children events passes through', () => {
    wrapper = mount(TodoList, { store, localVue })
    wrapper.findComponent(Todo).vm.$emit('updatetodo')
    wrapper.findComponent(Todo).vm.$emit('removetodo')
    expect(wrapper.emitted().updatetodo).toBeTruthy()
    expect(wrapper.emitted().removetodo).toBeTruthy()
  })
})
