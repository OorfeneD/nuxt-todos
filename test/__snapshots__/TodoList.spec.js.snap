// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component matchs no todos snapshot 1`] = `
"<ul class=\\"py-1\\">
  <li class=\\"text-sm text-gray-400 text-center py-2\\">
    There is nothing here. Add your first ToDo
  </li>
</ul>"
`;

exports[`TodoList Component matchs snapshot 1`] = `
"<ul class=\\"py-1\\">
  <todo-stub todo=\\"[object Object]\\"></todo-stub>
  <!---->
</ul>"
`;
